sw/device/silicon_creator/lib/drivers/lifecycle.c

```diff
@@ -0,0 +1,65 @@
+#include "sw/device/silicon_creator/lib/drivers/lifecycle.h"
+
+#include <assert.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "lc_ctrl_regs.h"
+
+// clang-format off
+const char * const lifecycle_state_name[] = {
+    "RAW",
+    "TEST_UNLOCKED0",
+    "TEST_LOCKED0",
+    "TEST_UNLOCKED1",
+    "TEST_LOCKED1",
+    "TEST_UNLOCKED2",
+    "TEST_LOCKED2",
+    "TEST_UNLOCKED3",
+    "DEV",
+    "PROD",
+    "PROD_END",
+    "RMA",
+    "SCRAP",
+    "POST_TRANSITION",
+    "ESCALATE",
+    "INVALID",
+};
+// clang-format on
+
+static_assert(ARRAYSIZE(lifecycle_state_name) == kLcStateNumStates,
+              "length of the lifecycle_state_name array doesn't match the "
+              "number of states.");
+
+#define LC_ASSERT(a, b) static_assert(a == b, "Bad value for " #a)
+LC_ASSERT(kLcStateRaw, LC_CTRL_LC_STATE_STATE_VALUE_RAW);
+LC_ASSERT(kLcStateTestUnlocked0, LC_CTRL_LC_STATE_STATE_VALUE_TEST_UNLOCKED0);
+LC_ASSERT(kLcStateTestLocked0, LC_CTRL_LC_STATE_STATE_VALUE_TEST_LOCKED0);
+LC_ASSERT(kLcStateTestUnlocked1, LC_CTRL_LC_STATE_STATE_VALUE_TEST_UNLOCKED1);
+LC_ASSERT(kLcStateTestLocked1, LC_CTRL_LC_STATE_STATE_VALUE_TEST_LOCKED1);
+LC_ASSERT(kLcStateTestUnlocked2, LC_CTRL_LC_STATE_STATE_VALUE_TEST_UNLOCKED2);
+LC_ASSERT(kLcStateTestLocked2, LC_CTRL_LC_STATE_STATE_VALUE_TEST_LOCKED2);
+LC_ASSERT(kLcStateTestUnlocked3, LC_CTRL_LC_STATE_STATE_VALUE_TEST_UNLOCKED3);
+LC_ASSERT(kLcStateDev, LC_CTRL_LC_STATE_STATE_VALUE_DEV);
+LC_ASSERT(kLcStateProd, LC_CTRL_LC_STATE_STATE_VALUE_PROD);
+LC_ASSERT(kLcStateProdEnd, LC_CTRL_LC_STATE_STATE_VALUE_PROD_END);
+LC_ASSERT(kLcStateRma, LC_CTRL_LC_STATE_STATE_VALUE_RMA);
+LC_ASSERT(kLcStateScrap, LC_CTRL_LC_STATE_STATE_VALUE_SCRAP);
+LC_ASSERT(kLcStatePostTransition, LC_CTRL_LC_STATE_STATE_VALUE_POST_TRANSITION);
+LC_ASSERT(kLcStateEscalate, LC_CTRL_LC_STATE_STATE_VALUE_ESCALATE);
+LC_ASSERT(kLcStateInvalid, LC_CTRL_LC_STATE_STATE_VALUE_INVALID);
+
+enum {
+  kBase = TOP_EARLGREY_LC_CTRL_BASE_ADDR,
+};
+
+lifecycle_state_t lifecycle_state_get(void) {
+  uint32_t value = bitfield_field32_read(
+      abs_mmio_read32(kBase + LC_CTRL_LC_STATE_REG_OFFSET),
```

Comment:
LC state is probably within the scope of sec_mmio, wdyt? We should add a TODO to integrate later.

Reply:
Filed #7026 and added a TODO.